### PR TITLE
Line fixes

### LINKED ISSUE #31: Investigate avoiding copying StateMachineRegisters when executing line number programs
From https://github.com/fitzgen/gimli/pull/27#issuecomment-238742174:

> It would be nice to avoid copying the registers for every line; it should be up to the caller to decide if a clone is needed. We could do that by storing the current LineNumberRow in StateMachine, and have execute() operate on the registers in that row, instead of having a separate copy. StateMachine::next() would return a reference to current LineNumberRow. Would need to find some way to defer the reset of discriminator/basic_block/prologue_end/epilogue_begin though.

I just looked into this a little and my conclusion is that we would need something like (wrapping?) `std::cell::Ref` but with a custom `Drop` impl to do the resetting.

Not going to dig any further at the moment.

### PR DESCRIPTION
**Pull request by philipc**
Fixes #31 

Not sure about a402366, happy to drop it if you prefer.

There's a few other fixes in the later patches too.  Some of them seem to have improved performance, but I didn't figure out which.  I've also been slack about adding tests for these... Can do it another day if you don't get to it first, not sure if you're currently working on this, so don't want to do too much that is going to conflict.

### CODE REVIEW

**Comment by fitzgen** [comment]
It looks like this isn't compiling yet?

```
error: no method named `push` found for type `&[u8]` in the current scope

    --> src/line.rs:1443:40

     |

1443 |         header.standard_opcode_lengths.push(0);

     |                                        ^^^^

error: no method named `push` found for type `&[u8]` in the current scope

    --> src/line.rs:1457:40

     |

1457 |         header.standard_opcode_lengths.push(1);

     |                                        ^^^^

error: no method named `push` found for type `&[u8]` in the current scope

    --> src/line.rs:1471:40

     |

1471 |         header.standard_opcode_lengths.push(3);

     |                                        ^^^^
```

**Comment by fitzgen** [comment]
Ah, you must have made this change in parallel with my new parse tests. I think you'll need to update them.

**Comment by fitzgen** [comment]
> I've also been slack about adding tests for these... Can do it another day if you don't get to it first, not sure if you're currently working on this, so don't want to do too much that is going to conflict.

Hold off on it -- I'm going to write (at least initial) tests for line number programs. In fact, I already have mostly complete unit tests for parsing, execution is next up.

In general, I think I should have written these tests before merging, but I got ahead of myself once I got the integration tests passing and the results seemed sane. I need to hold myself to a higher standard.

Going to give this PR a once over now...

File: src/line.rs [hunk 1]
```diff
@@ -63,197 +62,154 @@ impl<'input, 'header, Endian> StateMachine<'input, 'header, Endian>
...
-
-        self.registers.discriminator = 0;
-
-        row
+        let operation_advance = adjusted_opcode / self.row.header.line_range;
+        self.apply_operation_advance(operation_advance as u64);
     }
 
     /// Execute the given opcode, if a new row in the line number matrix is
     /// generated, return it.
```
Comment (fitzgen) [inline src/line.rs]: This comment needs to be updated. Also, not sure whether `Option<()>` is the best return type. Not sure if `bool` is any more satisfying. Maybe a custom `enum`? Can't really think of a good name for one...

File: src/line.rs [hunk 2]
```diff
@@ -277,14 +233,24 @@ impl<'input, 'header, Endian> StateMachine<'input, 'header, Endian>
...
-    pub fn next_row(&mut self) -> parser::ParseResult<Option<LineNumberRow<'input, 'header, Endian>>> {
+    pub fn next_row(&mut self)
+                    -> parser::ParseResult<Option<&LineNumberRow<'input, 'header, Endian>>> {
+        if self.row.end_sequence {
+            self.row.reset();
+        } else {
+            self.row.discriminator = 0;
+            self.row.basic_block = false;
+            self.row.prologue_end = false;
+            self.row.epilogue_begin = false;
```
Comment (fitzgen) [inline src/line.rs]: A comment here about Steps 3-7 from 6.5.2.1 would be nice. Is it valid to always apply this after yielding a row, not just for special opcodes?

**Comment by fitzgen** [comment]
> Not sure about a402366, happy to drop it if you prefer.

I'm not excited about this change, because I liked how the code matched 1:1 with the standard, but I don't feel strongly about it and it is still pretty obvious how the code matches the standard.

/me shrugs

File: src/line.rs (hunk 1 above)
Comment (philipc) [inline src/line.rs]: Yeah I wasn't sure either. I think `bool` is slightly better, so I changed it to that. I also made it private while I was at it.

File: src/line.rs (hunk 2 above)
Comment (philipc) [inline src/line.rs]: There's 3 opcodes that can yield a new row: EndSequence, Special, and Copy.  The reset behaviour is the same for Special and Copy. I've added comments.

**Comment by philipc** [comment]
> I'm not excited about this change, because I liked how the code matched 1:1 with the standard, but I don't feel strongly about it and it is still pretty obvious how the code matches the standard.

Dropped it then.

**Comment by philipc** [comment]
> Ah, you must have made this change in parallel with my new parse tests. I think you'll need to update them.

Yeah I hadn't noticed them. Should be all fixed now.

File: src/line.rs (hunk 2 above)
Comment (fitzgen) [inline src/line.rs]: Thanks for double checking!

File: src/line.rs [hunk 3]
```diff
@@ -277,14 +235,31 @@ impl<'input, 'header, Endian> StateMachine<'input, 'header, Endian>
...
+        // Perform any reset that was required after copying the previous row.
+        if self.row.registers.end_sequence {
+            // Previous opcode was EndSequence, so reset everything
+            // as specified in Section 6.2.5.3.
+            self.row.registers.reset(self.row.header.default_is_stmt);
+        } else {
+            // Previous opcode was one of:
+            // - Special - specified in Section 6.2.5.1, steps 4-7
+            // - Copy - specified in Section 6.2.5.2
+            // The reset behaviour is the same in both cases.
```
Comment (fitzgen) [inline src/line.rs]: Great comment, thanks!

**Comment by fitzgen** [comment]
Merged in 5aed77e1a8f1f37623dc4522e521ecb5e832e252 -- thanks!

**Pull request merged**